### ISSUE TITLE
Support deletion vector

### ISSUE THREAD

**Issue by wjones127** [delta-io/delta-rs#1094]
# Description

For protocol version 3, will want to support deletion vector. 

* [ ] Supporting reading with deletion vector
* [ ] Support delete operations using delete vector

https://github.com/delta-io/delta/blob/master/PROTOCOL.md#deletion-vectors

Question: how do we decide to rewrite vs use delete vector?

**Use Case**

This enables much faster deletes.

**Related Issue(s)**

Prerequisites:
 * #930 
 * #832

**Comment by houqp** [comment]
> Question: how do we decide to rewrite vs use delete vector?

This looks like a tradeoff between faster read performance v.s. faster write that need to be decided case by case? If so, might be better to just let the user decide depending on the expected workload pattern.

**Comment by guyrt** [comment]
+1 to supporting user-owned tradeoff decision. I'm investigating this feature internally and update patterns in individual tables likely dictate the right decision.

For instance, in many dimension tables, edits may be spread randomly through existing data and merge on read will be more efficient. For fact tables with mostly append pattern (but occasional fact updates), judicious partition plus copy on write may be superior.

**Comment by aersam** [comment]
Don't know if this helps, just tried to read a deletion vector file, and this seems to be working with the roaring crate:

```rust
use roaring::RoaringTreemap;
use std::io::Take;
use std::{fs::File, io::Read};

fn main() {
    let mut file =
        File::open("faker/deletion_vector_b3dfb787-c930-49c1-b78d-8333173f4db4.bin").unwrap();
    let mut buf = vec![0; 2];
    file.read(&mut buf).unwrap();
    let version = u16::from_le_bytes(buf.clone().try_into().unwrap());
    assert_eq!(version, 1);
    loop {
        let mut buf = vec![0; 3];
        let nrread = file.read(&mut buf).unwrap();
        if nrread == 0 {
            break;
        }

        let size_buf = [&[0], &buf[0..3]].concat();
        println!("size: {:?}", size_buf);
        let datasize = u32::from_be_bytes(size_buf.try_into().unwrap());
        println!("datasize: {}", datasize);
        if datasize == 0 {
            break;
        }
        let mut buf = vec![0; 4];
        file.read(&mut buf).unwrap();
        let magic = i32::from_le_bytes(buf.try_into().unwrap());
        assert!(magic == 1681511377);
        let take: Take<&File> = (&file).take(datasize as u64);
        let rdr = RoaringTreemap::deserialize_from(take).unwrap();
        for i in rdr.iter() {
            println!("index: {}", i);
        }
        let mut checksum_buf = vec![0; 4];
        println!("checksum_buf: {:?}", checksum_buf);
        file.read(&mut checksum_buf).unwrap();
    }
}

```

**Comment by aersam** [comment]
Would you accept a PR that does add the required metadata as a first step?

**Comment by roeap** [comment]
Hi @aersam - first of all thanks for the code snipplet, it actually samed me a bit of time working on this elsewhere.

In principle we always welcome contributions. In this case we also do, but there is one caveat. Elsewhere we are currently working hard on getting delta-kernel for rust released which will hopefully significantly boost our protocol support.

The more complex thing here is, that in order to support deletion vectors we have to either support reader V3 and writer v7 (i.e. table features), or support a whole bunch of other delta features as well.

Good news is we are actively working on it, but since this involves some larger blocks of work, its likely going to be a few weeks, before this can fully land...

With all that said, if you profit from having some intermediate partial support, I'd be happy to review PRs :)

**Comment by aersam** [comment]
Well if it's about weeks I can wait. I know that actually column mapping would be first, just thought that cannot be that hard ;) 

I did not know about delta-kernel for rust, I'm really glad to hear about it! To be honest I was a bit disappointed as I thought it will be in Java - nothing against Java, but I much prefer Rust, especially for embedding. Where do I find the code for delta-kernel/rust? Just to observe it a bit

Btw I also corrected the snipped, it had a bug when there are multiple vectors within file.

**Comment by alippai** [comment]
@roeap where can one follow the Delta kernel initiatives? I saw https://github.com/delta-io/delta/issues/1783 but that's not rust specific, right? Will it happen in this repo or will there be a delta-kernel-rs?

**Comment by djouallah** [comment]
fwiw; Fabric Datawarehouse just added support for deletion vectors and suddenly the delta table produced is no more compatible with Delta_rs :(

**Comment by boccileonardo** [comment]
Is this feature still on the roadmap? Tables produced by recent databricks runtime include deletion vectors by default, so it seems to me that reading them through rust-based solutions like polars is not currently possible natively.

**Comment by dylan-lee94** [comment]
Running into the same issue, the latest databricks runtime have deletion vectors enabled by default and our admin won't turn it off. This breaks our python code that is reading with DeltaTable or polars.

**Comment by djouallah** [comment]
> Running into the same issue, the latest databricks runtime have deletion vectors enabled by default and our admin won't turn it off. This breaks our python code that is reading with DeltaTable or polars.

as a temporary workaround, duckdb do support reading delta table with deletion vectors using the delta extension based on delta kernel not delta_rs

**Comment by 1ct0** [comment]
I am writing to add support in the request for this enhancement. Databricks now enables deletion vectors on tables by default when creating a new table using a SQL warehouses or Databricks Runtime 14.1 or above. Interacting with these tables (both reading and writing records) using polars, duckdb, etc. is not possible at the moment due to those libraries' reliance on delta-rs.

Is this still on the roadmap? Or is anyone aware of a workaround?

**Comment by roeap** [comment]
I am currently working on integrating more delta-kernel-rs functinoalities, that are currently handled via custom implememntations here.

Kernel already supports deletion vectors. This is a significant amount of work, but we are also making progres ...

So in short, yes, this very much planned, but currently need to figure out a bit more to give a realistic estiamte when we can enable this.

**Comment by choucavalier** [comment]
@roeap hey! any work going on regarding **write** support for deletion vectors?

**Comment by PeterThramkrongart** [comment]
I'll bump that. My team, needs to decide on a deletion scheme. Can we expect deletion vectors to be supported somewhat soon?

**Comment by mbelak-dtml** [comment]
+1 on this. Being able to read tables with deletion vectors would be awesome.

**Comment by pall-j** [comment]
Hi, I wanted to follow up on this issue. Could you please provide an update on the status? This is currently a major blocker for our project, and any information on timeline or next steps would be greatly appreciated.

**Comment by mwc360** [comment]
Any updates on DV support? DV use is common in Spark scenarios so it unfortunately really kills the interop story that the many other engines dependent on Delta-rs still can't interact with DV enabled tables :/

**Comment by gz** [comment]
I agree with @mwc360, most delta-tables out there contain deletion vectors by now which, unfortunately, makes this crate not very useful for reading them. Is there an ETA on support for deletion vector?

**Comment by djouallah** [comment]
@gz as a workaround, you can use duckdb to read delta, it does support deletion vectors and column mapping

**Comment by ion-elgreco** [comment]
@pall-j @mwc360 @gz this is a community driven project, so unless someone rolls up their sleeves and builds it you will get it but there is no point in asking for updates

**Comment by edgBR** [comment]
Hi @ion-elgreco 

Just for my understanding what will be needed for this?

I guess we need to update the reader and the writer?

Do you know if this is the latest spec:

https://docs.google.com/document/d/1lv35ZPfioopBbzQ7zT82LOev7qV7x4YNLkMr2-L5E_M/edit?pli=1&tab=t.0#heading=h.dd2cc57oc5wk

**Comment by ion-elgreco** [comment]
@edgBR yes both require an update to handle deletion vectors. Kernel-rs already has this but it's not integrated yet

**Comment by edgBR** [comment]
Hi @ion-elgreco 

As you know Im not that well versed with this project.

I was reviewing the code and I see that you have:

https://github.com/delta-io/delta-rs/tree/284608107f49a62ac3655567837345d3d7ad7749/crates/core/src/kernel

When you say the kernel-rs is ready for that are you refering to the code above or to the official delta-kernel-rs on:

https://github.com/delta-io/delta-kernel-rs

**Comment by edgBR** [comment]
Hi,

I think my question above is stupid (sorry I dont know rust).

I assume that you are importing the delta-kernel-rs in the cargo file with:


```
[workspace.dependencies]
delta_kernel = { version = "0.13.0", features = [
    "arrow-55",
    "default-engine-rustls",
    "internal-api",
]}
```

I assume then we need to start by adding:

`use delta_kernel::actions::deletion_vector::DeletionVectorDescriptor;`

Somewhere?

I am willing to learn and contribute on this, any directions are appreciated.

**Comment by getChan** [comment]
I successfully read the deletion vector table today.
Has the feature from delta-kernel-rs been integrated into delta-rs? @roeap

**Comment by choucavalier** [comment]
@getChan reading deletion vectors has been implemented in delta-rs for a while now

what's missing now is writing deletion vectors

right now, if you make a small modification to a row in a delta table (e.g. remove 3 rows, or modifying a value in a row) it's gonna rewrite the underlying parquet file entirely. if you do that many times in a row, you end up with 50GB of storage for 30MB of data

that's not the scenario delta tables are meant to be used for though. but deletion vectors were designed to make that scenario a little bit less dramatic

**Comment by boccileonardo** [comment]
@getChan @choucavalier When you say read support is already there, do you mean that tools like polars should already support reading tables with enabled deletion vectors? This does not currently seem to be the case.

Based on the stack trace, I actually assumed it was still deltalake package that was returning this error:
```
deltalake/table.py:166, in DeltaTable.__init__(self, table_uri, version, storage_options, without_files, log_buffer_size)
CommitFailedError: Unsupported reader features required: [DeletionVectors]
```

**Comment by getChan** [comment]
@boccileonardo 

My successful read version was **delta-rs 0.28.1**. I used the built-in DataFusion as the query engine.
I was also able to successfully read deletion vector tables with the latest DuckDB (1.3.2).

**Comment by boccileonardo** [comment]
I'm on the latest (1.1.4) python distribution of deltalake when encountering the error above, so either this functionality is only available on the rust crate and not on the python distribution, or there's an issue with the polars implementation.

**Comment by ion-elgreco** [comment]
@getChan that's a bug, you shouldnt be able to read/write tables with deletion vectors

**Comment by atzannes** [comment]
+1 to adding support for reading delta tables with deletion vectors. This is a pretty important feature for us because we have a Databricks environment that by default enables deletion vectors, but we want our data to be easily accessible outside of the Databricks environment using Polars (which uses delta-rs unless I'm mistaken) for quick analytics. Right now we're making due by using DuckDB to load the data locally and then move it into Polars, but it would be much nicer to be able to directly interact with these tables using Polars.

**Comment by kdn36** [comment]
Polars OSS would like to support reading (and eventually writing) files with deletion vectors directly. To that end, we would like to see deletion vectors exposed through the Python bindings. Is that in the plan / can that be arranged? Thanks in advance.

xref https://github.com/pola-rs/polars/issues/26369

**Comment by rtyler** [comment]
This ticket is kind of a monolith. We have preliminary support for Deletion Vectors through the great work that @roeap has been doing and subsequent requests should be pushed into new issues.

**Comment by kdn36** [comment]
> +1 to adding support for reading delta tables with deletion vectors. This is a pretty important feature for us because we have a Databricks environment that by default enables deletion vectors, but we want our data to be easily accessible outside of the Databricks environment using Polars (which uses delta-rs unless I'm mistaken) for quick analytics. Right now we're making due by using DuckDB to load the data locally and then move it into Polars, but it would be much nicer to be able to directly interact with these tables using Polars.

To close the loop: PR https://github.com/pola-rs/polars/pull/26867 has been merged using the delta-rs work. It will be part of the next polars release. Testing and feedback appreciated (in the polars GH repo, unless it is a delta-rs issue). Big thanks to the delta-rs team.
